File: src/storeFx/fetchArtists.ts
import { deepEqual } from '@tanstack/react-router';

import * as Fx from '../fx/fx';
import { mergeIntoMap } from '../mergeIntoMap/mergeIntoMap';
import type { Store } from '../store/types';
import type { BaseArtist } from '../subsonic';
import { getArtists } from '../subsonic';

export function fetchArtists() {
  return Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) =>
    Fx.bind(
      Fx.provide(getArtists(), {
        credentials: store.getState().auth.credentials,
      }),
      ({ index }) => {
        const artists: BaseArtist[] = [];
        for (const { artist } of index) artists.push(...artist);

        const state = store.getState();
        const byId = mergeIntoMap(state.artists.byId, artists, x => x.id);
        const listIds = artists.map(artist => artist.id);

        if (
          byId !== state.artists.byId ||
          !deepEqual(listIds, state.artists.listIds)
        ) {
          store.setState(prevState => ({
            artists: { ...prevState.artists, byId, listIds },
          }));
        }

        return Fx.Ok();
      },
    ),
  );
}

File: src/storeFx/fetchArtists.ml
open Fx
open Monad_syntax
open MergeIntoMap
open Zustand

let make () =
  let* Deps.{ store } = ask () in
  let* { index } =
    let state = getState store in
    Subsonic.GetArtists.make ()
    |. provide ~deps:{ credentials = state.auth.credentials }
  in
  let artists =
    index
    |. Js.Array.reduce ~init:[||] ~f:(fun acc entry ->
           acc
           |> Js.Array.pushMany ~values:Subsonic.Artists.(entry.artist)
           |> ignore;
           acc)
  in
  let state = getState store in
  let byId = mergeIntoMap state.artists.byId artists (fun x -> x.id) in
  let listIds =
    Some
      (artists
      |> Js.Array.map ~f:(fun artist -> Subsonic.BaseArtist.(artist.id)))
  in
  if
    byId != state.artists.byId
    || not (Router.deepEqual listIds state.artists.listIds)
  then
    store
    |. setState (fun prevState ->
           { prevState with artists = { prevState.artists with byId; listIds } });
  Ok ()
